File: src/app/core/network.ts
/**
 * Whether this device is on a connection that charges by the byte.
 *
 * Two sources, because no single one covers the platforms this app runs on:
 *
 *   - `navigator.connection`, the Network Information API. Chromium only,
 *     which in practice means the Android build and a desktop browser. It is
 *     the better answer where it exists: the OS tells the engine directly, and
 *     it fires a `change` event instead of being polled.
 *   - `network_kind`, the native command, which walks the interface list. This
 *     is the ONLY answer on iOS, where WKWebView ships no Network Information
 *     API at all - and iOS is precisely where mobile data is a real bill.
 *
 * And a third state that is not a source: `unknown`. A browser tab, a desktop
 * build on Windows, an older native binary without the command, a device
 * mid-handover. Every caller must treat unknown as "do not block", because a
 * download that silently stops is a worse failure than one that costs a few
 * megabytes - the first reads as the feature being broken and gives nobody a
 * clue why, the second shows up somewhere a person can act on it.
 */

import { isTauri, tauriCall } from './tauri.ts';

export type NetworkKind = 'wifi' | 'cellular' | 'unknown';

/**
 * The shape of `navigator.connection` we actually read.
 *
 * Typed here rather than pulled from lib.dom: the Network Information API is
 * not in TypeScript's DOM types, because it has never been more than a draft
 * that one engine shipped.
 */
interface NetworkInformation extends EventTarget {
  /** 'wifi' | 'cellular' | 'ethernet' | 'none' | 'unknown' | ... */
  type?: string;
}

function connection(): NetworkInformation | null {
  if (typeof navigator === 'undefined') return null;
  const nav = navigator as Navigator & { connection?: NetworkInformation };
  return nav.connection ?? null;
}

/** The web answer, where the engine has one worth reading. */
function fromWeb(): NetworkKind {
  const type = connection()?.type;
  if (type === 'cellular') return 'cellular';
  // Ethernet counts as Wi-Fi here for the same reason the native side folds
  // them together: the switch is about the bill, not the radio.
  if (type === 'wifi' || type === 'ethernet') return 'wifi';
  // 'none', 'unknown', 'other', 'mixed', or the property missing entirely.
  // None of those is evidence, and 'none' in particular must not read as
  // cellular - a device with no network needs no protecting from a download
  // that cannot start anyway.
  return 'unknown';
}

/**
 * The last native answer, refreshed rather than awaited at the point of use.
 *
 * The native probe is one IPC round trip over an interface walk - cheap, but
 * not free, and not synchronous. Callers that must be right (the sweep) await
 * a fresh one; callers that only draw (the settings row) read this.
 */
let nativeKind: NetworkKind = 'unknown';
let probing: Promise<NetworkKind> | null = null;

async function fromNative(): Promise<NetworkKind> {
  if (!isTauri()) return 'unknown';
  // Single-flight: the settings pane mounting while a sweep is deciding
  // should be one interface walk, not two.
  probing ??= tauriCall<string>('network_kind')
    .then((answer) => {
      // A null is an older binary without the command, and reads as unknown -
      // the contract tauriCall exists for.
      const kind: NetworkKind =
        answer === 'wifi' || answer === 'cellular' ? answer : 'unknown';
      nativeKind = kind;
      return kind;
    })
    .catch(() => 'unknown' as NetworkKind)
    .finally(() => {
      probing = null;
    });
  return probing;
}

/**
 * What this device is on, freshly asked.
 *
 * The web answer wins when it is definite: it comes from the engine's own
 * connectivity monitor rather than a name-matching heuristic, so where both
 * can speak it is the more trustworthy of the two.
 */
export async function networkKind(): Promise<NetworkKind> {
  const web = fromWeb();
  if (web !== 'unknown') return web;
  return fromNative();
}

/** The last known answer, without waiting. For drawing, not for deciding. */
export function networkKindNow(): NetworkKind {
  const web = fromWeb();
  return web !== 'unknown' ? web : nativeKind;
}

/**
 * Whether downloading right now would spend somebody's mobile data.
 *
 * Deliberately NOT `kind !== 'wifi'`. Unknown has to fall on the permissive
 * side, and writing the test this way makes that visible at every call site
 * instead of hiding it behind an inverted comparison.
 */
export async function onMeteredConnection(): Promise<boolean> {
  return (await networkKind()) === 'cellular';
}

// --- staying current --------------------------------------------------------

const watchers = new Set<(kind: NetworkKind) => void>();

function announce(kind: NetworkKind): void {
  for (const w of watchers) w(kind);
}

async function recheck(): Promise<void> {
  const before = networkKindNow();
  // Always re-probe natively rather than trusting the cache: this runs on the
  // events that mean the connection may have just changed, which is exactly
  // when the cached value is the stale one.
  await fromNative();
  const after = networkKindNow();
  if (after !== before) announce(after);
}

let wired = false;

/**
 * Watch for the connection changing kind.
 *
 * Three signals, none of which is reliable alone: the Network Information
 * API's own `change` (Chromium only), `online` (fires on a reconnect, not on a
 * Wi-Fi-to-cellular handover), and coming back to the foreground - which is
 * the one that actually catches walking out of the house with the app open,
 * since the webview is frozen while the phone is in a pocket.
 */
export function onNetworkChange(fn: (kind: NetworkKind) => void): () => void {
  watchers.add(fn);

  if (!wired && typeof window !== 'undefined') {
    wired = true;
    const bump = () => void recheck();
    connection()?.addEventListener('change', bump);
    window.addEventListener('online', bump);
    document.addEventListener('visibilitychange', () => {
      if (!document.hidden) bump();
    });
    // Prime it, so the first reader is not looking at the 'unknown' this
    // module starts life holding.
    bump();
  }

  return () => {
    watchers.delete(fn);
  };
}

File: src-tauri/src/network.rs
//! What kind of network this device is on, for "only download on Wi-Fi".
//!
//! WHY THIS IS NATIVE AT ALL. The web platform already answers this, and on
//! Android it answers it well: `navigator.connection.type` is a Chromium
//! feature and the Android build is Chromium. iOS is the problem. WKWebView
//! ships no Network Information API whatsoever, so the one device where mobile
//! data is an actual bill is the one device the web answer cannot reach. A
//! Wi-Fi-only switch that silently does nothing on iPhone would be worse than
//! no switch, because it would read as a promise.
//!
//! So this asks the OS through the interface list. It is a heuristic and not a
//! route-table lookup: the question being asked is "would bytes cost money",
//! which is about the BILL rather than the radio, and the interface names
//! answer that well enough on every platform that has a cellular modem. Where
//! it cannot tell, it says so - see `network_kind`.

/// What this device is connected through, as far as the interface list shows.
///
/// Three answers, and `unknown` is a real one rather than a failure. A build
/// that cannot look (Windows), a device mid-handover, and a machine with no
/// network at all all land there. Every caller treats it as "do not block",
/// because a download that silently stops is a worse bug than one that costs a
/// few megabytes: the first looks like the feature is broken and gives no clue
/// why, the second is visible on a bill the user can act on.
#[tauri::command]
pub fn network_kind() -> &'static str {
    probe()
}

#[cfg(unix)]
fn probe() -> &'static str {
    use std::ffi::CStr;

    let mut list: *mut libc::ifaddrs = std::ptr::null_mut();
    // SAFETY: getifaddrs writes an owned linked list into `list` on success and
    // leaves it alone otherwise; the freeifaddrs below releases it exactly once.
    if unsafe { libc::getifaddrs(&mut list) } != 0 {
        return "unknown";
    }

    let mut wifi = false;
    let mut cellular = false;
    let mut node = list;
    while !node.is_null() {
        // SAFETY: node is non-null here and came from getifaddrs, which
        // guarantees a well-formed entry with an ifa_next terminating in null.
        let ifa = unsafe { &*node };
        node = ifa.ifa_next;

        if ifa.ifa_addr.is_null() || ifa.ifa_name.is_null() {
            continue;
        }
        let flags = ifa.ifa_flags as i32;
        // UP is "configured", RUNNING is "actually has a carrier". Both, or a
        // laptop with Wi-Fi enabled and no network joined reports Wi-Fi.
        if flags & libc::IFF_UP == 0 || flags & libc::IFF_RUNNING == 0 {
            continue;
        }
        if flags & libc::IFF_LOOPBACK != 0 {
            continue;
        }
        // SAFETY: ifa_addr is non-null (checked) and points at a sockaddr whose
        // family tag says which concrete layout follows.
        if !unsafe { routable(ifa.ifa_addr) } {
            continue;
        }

        // SAFETY: ifa_name is non-null (checked) and NUL-terminated.
        let name = unsafe { CStr::from_ptr(ifa.ifa_name) }.to_string_lossy();
        if is_cellular(&name) {
            cellular = true;
        } else if is_unmetered(&name) {
            wifi = true;
        }
    }

    // SAFETY: `list` came from a successful getifaddrs and is freed once, after
    // the walk above has finished reading every node.
    unsafe { libc::freeifaddrs(list) };

    // Wi-Fi wins when both are up, and on iOS both being up is the NORMAL
    // state rather than the exception: the cellular interface keeps its
    // address while you are on Wi-Fi. Reading "cellular" from its mere
    // presence would report cellular to a phone sitting on a home network,
    // which is the failure mode that makes this kind of check useless.
    if wifi {
        "wifi"
    } else if cellular {
        "cellular"
    } else {
        "unknown"
    }
}

/// Windows has no getifaddrs, and a Windows desktop is not where a mobile data
/// bill comes from. Unknown, which reads as "do not block".
#[cfg(not(unix))]
fn probe() -> &'static str {
    "unknown"
}

/// Whether an address could carry traffic off this machine.
///
/// The link-local checks are what let an idle Wi-Fi interface read as absent.
/// A phone with Wi-Fi switched on but nothing joined still has `en0` up and
/// running carrying an `fe80::` address, and counting that would report Wi-Fi
/// for a device doing every byte of its work over cellular - the exact case
/// this whole module exists to catch.
///
/// # Safety
/// `addr` must be non-null and point at a `sockaddr` whose `sa_family` field
/// correctly describes the layout that follows it.
#[cfg(unix)]
unsafe fn routable(addr: *const libc::sockaddr) -> bool {
    match unsafe { (*addr).sa_family } as i32 {
        libc::AF_INET => {
            let v4 = unsafe { &*(addr as *const libc::sockaddr_in) };
            // 169.254.0.0/16 - what a machine assigns itself when no DHCP
            // server ever answered.
            u32::from_be(v4.sin_addr.s_addr) >> 16 != 0xA9FE
        }
        libc::AF_INET6 => {
            let v6 = unsafe { &*(addr as *const libc::sockaddr_in6) };
            let b = v6.sin6_addr.s6_addr;
            // fe80::/10
            !(b[0] == 0xFE && b[1] & 0xC0 == 0x80)
        }
        // A link-layer entry (AF_LINK/AF_PACKET) says nothing about whether the
        // interface can reach anything, so it is not evidence either way.
        _ => false,
    }
}

/// Interface names that mean "these bytes are metered".
#[cfg(unix)]
fn is_cellular(name: &str) -> bool {
    // iOS names every cellular context pdp_ipN. Android and desktop Linux use
    // whatever the modem driver calls itself, which is vendor-specific: rmnet
    // is Qualcomm, ccmni MediaTek, and wwan/qmi the generic USB modem stack.
    name.starts_with("pdp_ip")
        || name.starts_with("rmnet")
        || name.starts_with("ccmni")
        || name.starts_with("wwan")
        || name.starts_with("qmi")
}

/// Interface names we treat as costing nothing per byte.
#[cfg(unix)]
fn is_unmetered(name: &str) -> bool {
    // `en*` covers both Wi-Fi and Ethernet on Apple platforms, which is right
    // rather than sloppy - the question is the bill, not the radio.
    //
    // Deliberately NOT a catch-all, and this is the half that has to stay
    // narrow: awdl0 (AirDrop), utun* (VPN), ap1 (personal hotspot, where the
    // bytes leaving the device are somebody's cellular) and bridge* all reach
    // this function while a phone is tethering, and every one of them must
    // fail it. A VPN is the subtle case: traffic routes over utun*, but the
    // interface underneath is still en0 or pdp_ip0, so ignoring the tunnel is
    // what keeps the answer correct.
    name.starts_with("en")
        || name.starts_with("wlan")
        || name.starts_with("eth")
        || name.starts_with("wl")
}

#[cfg(all(test, unix))]
mod tests {
    use super::*;

    #[test]
    fn tunnels_and_hotspots_are_not_wifi() {
        // The names that are up and running on a TETHERING iPhone. If any of
        // these counted, the phone sharing its cellular connection would
        // report Wi-Fi and download over exactly the link being paid for.
        for name in ["utun0", "utun3", "awdl0", "ap1", "bridge100", "llw0"] {
            assert!(!is_unmetered(name), "{name} must not count as unmetered");
            assert!(!is_cellular(name), "{name} is not a modem either");
        }
    }

    #[test]
    fn the_modem_names_are_recognised() {
        for name in ["pdp_ip0", "pdp_ip1", "rmnet_data0", "ccmni0", "wwan0"] {
            assert!(is_cellular(name), "{name} is a modem");
            assert!(!is_unmetered(name), "{name} must never read as unmetered");
        }
    }

    #[test]
    fn ordinary_interfaces_still_read_as_unmetered() {
        for name in ["en0", "en1", "eth0", "wlan0"] {
            assert!(is_unmetered(name), "{name} should count as unmetered");
        }
    }

    #[test]
    fn probing_this_machine_answers_one_of_three() {
        // Not asserting WHICH - it depends on the machine running the tests -
        // only that the walk completes, frees its list and returns a value the
        // frontend's union actually contains.
        assert!(matches!(probe(), "wifi" | "cellular" | "unknown"));
    }
}

